designs/2023-only-run-reporting-rules/README.md

```diff
@@ -0,0 +1,153 @@
+- Repo: eslint/eslint
+- Start Date: 2023-01-16
+- RFC PR: <https://github.com/eslint/rfcs/pull/104>
+- Authors: Maddy Miller
+
+# Only run reporting lint rules
+
+## Summary
+
+Currently, ESLint will run all rules that are not marked as `off` in the configuration.
+This RFC proposes adding a way to configure which rules are actually run, to reduce linting
+time and better match the reporting outcome. Currently, when a rule is marked as `warn`,
+ESLint will still run the rule but not report the results when run under `--quiet`. The
+intended outcome of this RFC is to allow users to not run these rules when unnecessary.
+
+## Motivation
+
+A common use-case of the `warn` level is to report errors to the developer in their IDE,
+but to completely ignore it when running ESLint on the command line via the `--quiet` flag.
+When running with many rules set to the `warn` level, this can cause a significant amount
+of overhead on ESLint runtime. On an example large codebase with 154988 lines of TypeScript
+code, I found that by disabling all rules set to warn, I was able to reduce ESLint runtime
+from 4 minutes and 39 seconds to just 22 seconds. As the `warn` rules are entirely ignored
+in this situation, this is a substantial amount of wasted time and resources that could be
+avoided.
+
+## Detailed Design
+
+The proposed implementation is to modify the `--quiet` flag so that ESLint will skip
+running any rules that are set to the `warn` level.
+
+From an API perspective, this would be implemented by a filter function that filters down to
+which rules should be run. The function would take a list of the rule configuration `(ruleId, severity)`,
+and return the rule that should be run. For simplicity, this function should not be
+given rules marked as `off`, as if this function handled existing behavior then users of the
+API would have to mimic that when attempting to extend it.
+
+In `cli.js`'s `translateOptions` function, the `filterRules` option should be assigned to
+a function that filters to rules with a `severity` of 2 (error) when the `--quiet` flag is applied, otherwise
+acts as an identity function. In `eslint/flat-eslint.js`, the `filterRules` should be taken from
+the `eslintOptions` object, and passed down to the `linter.verifyAndFix` call.
+
+Within `linter.js`, the API should be added to `VerifyOptions`, and will be passed down into and
+utilized within the `runRules` function before the current `configuredRules` loop. Rather than the
+current configuredRules loop, this should extract the `severity` and `rule` existence checks and
+build a list of `{ ruleId, severity }` objects. This new list should be passed to `filterRules`,
+and the resulting list should be iterated on instead of `configuredRules`.
```

Comment:
```suggestion
Within `linter.js`, the API should be added to `VerifyOptions`, and will be passed down into and
utilized within the `runRules` function before the current `configuredRules` loop. Rather than the
current configuredRules loop, this should use `ruleFilter` as a filter to check the `severity` and `ruleId` to
build a list of `ruleId`s. This new list should be iterated on instead of `configuredRules`.
```